### PR TITLE
obj_integrity moved up to atom level, hulk wallsmashing reworked, shoot-downable walls

### PR DESCRIPTION
**Pull request by PKPenguin321**
## About The Pull Request

- obj_integrity has been renamed to atom_integrity.
- obj_integrity and related vars have been moved to /atom/ from /obj/
- Integrity-related procs have also been moved in the same way
- A handful of said procs have been renamed ("obj" mentions in proc names changed to "atom")
- Reinforced walls now have 1000 max integrity, twice that of a regular wall (as well as high armor and damage deflection, making them not-easy to shoot down without penetrating rounds)
- Guns now damage walls
- Hulks now deal 100 integrity damage to walls per hit, rather than having a random chance to break walls
- Requested by @Rohesie 
- Closes https://github.com/tgstation/tgstation/pull/51389

## Why It's Good For The Game

- More generalized systems like this = cooler potential for interactions (integrity as a system is very highly configurable and has features like things breaking before being destroyed)
- Hulk nerfed and made consistent
- You can line up 50 emitters and bore a hole through the station very slowly

## Changelog
:cl:
code: Object integrity system has been moved up to the atom level
add: Functionally at the moment this means that walls can be damaged and eventually destroyed by projectiles
balance: Hulks now damage walls instead of randomly destroying them, and will take 5 hits to break down a wall, and 10 for an r-wall
/:cl:

### CODE REVIEW

**Comment by PKPenguin321** [comment]
Don't be scared by the massive amount of files changed, that's from renaming vars/procs. Bulk of code changed is in atoms.dm, walls.dm, objs.dm, obj_defense.dm, projectile.dm

File: code/game/turfs/closed/walls.dm [hunk 1]
```diff
@@ -37,6 +37,16 @@
 	if(is_station_level(z))
 		GLOB.station_turfs += src
 
+	//integrity stuff
```
Comment (PKPenguin321) [inline code/game/turfs/closed/walls.dm:40]: This is copypasted from atom initialize. I think walls are deployed before initializing an atom works properly, so if this isn't here, walls aren't initialized with armor/atom_integrity being set to max_integrity. This is a hack and I'd like if there was a cleaner way to do this.

**Comment by spookydonut** [comment]
mobs have integrity now?

**Comment by PKPenguin321** [comment]
Yeah, functionally does nothing though

**Comment by spookydonut** [comment]
lighting objects have integrity now?

**Comment by Rohesie** [comment]
I believe a change to make the system more generic and unified is very positive.
There's two ways of accomplishing this:

1) The way this PR does it, moving the generic concept of damage and resistance back to the atom level. You could argue this concept doesn't apply  to things such as visual objects or other abstract ones such as effects, but as objects they already have these variables, so that'd only apply for a few movable atoms.

2) Making only the damage procs universal and using intermediary procs such as `get_max_integrity()` and `get_current_integrity()` which fetch the relevant variables if any.

I favor the first method, but both work for me.
Thoughts?

**Comment by Shadowflame909** [comment]
How does this affect the standard holoparasite which is another wall smasher.

Can swarmers now damage the walls integrity?

File: code/game/turfs/closed/walls.dm (hunk 1 above)
Comment (nemvar) [reply code/game/turfs/closed/walls.dm:40]: turf/Initialize doesn't call parent. That's why you are seeing this behaviour.

Comment (nemvar) [reply code/game/turfs/closed/walls.dm:40]: Can you at least encapsulate this entire block in another proc and then call that from both turf initialize and atom initialize? Also this currently results in dense turfs that don't inherit from /wall to have 0 integrity.

**Comment by Tlaltecuhtli** [comment]
👍d and 💳d , waiting for 10 emitters cannon to cut down the whole station

**Comment by Ghilker** [comment]
how this will affect mobs? do mobs have integrity too now?

**Comment by PKPenguin321** [comment]
>How does this affect the standard holoparasite which is another wall smasher.
Can swarmers now damage the walls integrity?

This PR presently affects neither of those things, though either could be easily changed to use this system.

File: code/game/turfs/closed/walls.dm (hunk 1 above)
Comment (PKPenguin321) [reply code/game/turfs/closed/walls.dm]: Thanks, this helped

**Comment by PKPenguin321** [comment]
Oh, and also, damage deflection now applies to any damage type, not just melee damage. Not certain of the balance implications of this but I don't think it actually affects anything except for bullet damage on walls as added in this PR.

**Comment by Qustinnus** [comment]
@spookydonut lighting objects had it already anyways seeing as its obj integrity atm anyways

**Comment by Qustinnus** [comment]
also this should be testmerged over a long period because I dont trust this to not cause bugs.

File: code/game/turfs/closed/walls.dm [hunk 2]
```diff
@@ -14,6 +14,9 @@
 
 	flags_ricochet = RICOCHET_HARD
 
+	armor = list("melee" = 95, "bullet" = 25, "laser" = 25, "energy" = 100, "bomb" = 0, "bio" = 100, "rad" = 0, "fire" = 95, "acid" = 0)
```
Comment (Rohesie) [inline code/game/turfs/closed/walls.dm:17]: Walls may warrant higher bullet and laser armor.

File: code/game/turfs/closed/wall/reinf_walls.dm [hunk 3]
```diff
@@ -6,6 +6,10 @@
 	opacity = 1
 	density = TRUE
 
+	max_integrity = 1000 //2x the integrity of a plain wall
```
Comment (Rohesie) [inline code/game/turfs/closed/wall/reinf_walls.dm:9]: I'd personally go higher, to 1500 or so.

**Comment by Rohesie** [comment]
> 
> also this should be testmerged over a long period because I dont trust this to not cause bugs.

The actual changes are very few.  This is mostly renaming mass-replacements.

**Comment by Qustinnus** [comment]
the changes being few does not mean that adding integrity to things like turfs will not have enormous implications.

**Review by ghost — CHANGES_REQUESTED** [review]
1. space and other supposed to be indestructible turfs dont have the right resistance flags.
2. You need to move the things that affect health (e.g. take_damage) actually run on atom level rather than running on obj level as they do now. e.g.
```
/obj/attacked_by(obj/item/I, mob/living/user)
	if(I.force)
		user.visible_message("<span class='danger'>[user] hits [src] with [I]!</span>", \
					"<span class='danger'>You hit [src] with [I]!</span>", null, COMBAT_MESSAGE_RANGE)
		//only witnesses close by and the victim see a hit message.
		log_combat(user, src, "attacked", I)
	take_damage(I.force, I.damtype, "melee", 1)
```
you've moved the procs up to atom level but most of the procs are still being run in a different level. If this is done you also need to go through attacked_by procs on turfs in general and make sure they still work.

And this one is just an example, there's a ton of cases where a proc typecasts to obj level and then makes it take damage, all of this needs to be moved to atom instead.

File: code/game/turfs/closed/walls.dm [hunk 4]
```diff
@@ -138,16 +144,14 @@
...
-					"<span class='danger'>You smash \the [src]!</span>", \
-					"<span class='hear'>You hear a booming smash!</span>")
+	playsound(src, 'sound/effects/meteorimpact.ogg', 100, TRUE)
+	user.say(pick(";RAAAAAAAARGH!", ";HNNNNNNNNNGGGGGGH!", ";GWAAAAAAAARRRHHH!", "NNNNNNNNGGGGGGGGHH!", ";AAAAAAARRRGH!" ), forced = "hulk")
+	playsound(src, 'sound/effects/bang.ogg', 50, TRUE)
+	add_dent(WALL_DENT_HIT)
+	user.visible_message("<span class='danger'>[user] smashes \the [src]!</span>", \
+				"<span class='danger'>You smash \the [src]!</span>", \
+				"<span class='hear'>You hear a booming smash!</span>")
+	take_damage(100, BRUTE, "melee", FALSE, user.dir, 100)
```
Comment (Qustinnus) [inline code/game/turfs/closed/walls.dm:154]:
```suggestion
	take_damage(hulk_damage(), BRUTE, "melee", FALSE, user.dir, 100)
```

Comment (Qustinnus) [reply code/game/turfs/closed/walls.dm:154]: and again, not a fan of us having 2 procs on turf and obj that basically do the same thing.

Comment (PKPenguin321) [reply code/game/turfs/closed/walls.dm:154]: this specific proc is really weird, all it does is return 150. might as well make it a define or not even use it

Comment (Qustinnus) [reply code/game/turfs/closed/walls.dm:154]: I don't mind just make sure that both obj and Turf uses the same method and value.

Comment (Rohesie) [reply code/game/turfs/closed/walls.dm:154]: A define would be okay.

**Comment by ExcessiveUseOfCobblestone** [comment]
Can you have hulk scale with punch damage on species? I would rather it be a (sensible) range than a strict 5 for walls, 10 for r-walls.

**Comment by TemporalOroboros** [comment]
Only humans can have hulk last I checked

**Comment by ExcessiveUseOfCobblestone** [comment]
> 
> 
> Only humans can have hulk last I checked

You're correct, but it would be nice futureproofing (as well as provide some range rather than a flat number each time).

**Comment by PKPenguin321** [comment]
You mean you want things like mothhulks to be weaker?

**Comment by PKPenguin321** [comment]
i'm thinking things with the indestructible flag should no longer even be attackable, so as to stop people from attacking space. thoughts?

**Comment by necromanceranne** [comment]
Moths do not have weaker punches. All roundstart species have the same punch values. I think only golems have a very diverse range of punch values, and they can't at the moment become a hulk.

**Comment by Rohesie** [comment]
> 
> 
> i'm thinking things with the indestructible flag should no longer even be attackable, so as to stop people from attacking space. thoughts?

Yep, that's not a bad idea.

**Comment by ExcessiveUseOfCobblestone** [comment]
> 
> 
> You mean you want things like mothhulks to be weaker?

if they had lower punch values and could obtain hulk, yes. Even if there can only be humanhulks I think this would be nice since there may be items later that edited your punch strength, and this should affect your ability as hulk as well (imo)

Instead of the strict number you would just have a random number between the punch range for the species (probably physiology instead) converted into your desired obj integrity.

**Comment by spookydonut** [comment]
Is this stale?

**Comment by PKPenguin321** [comment]
Yeah sorry about that, been distracted

**Comment by PKPenguin321** [comment]
sorry for the letdown but i'm gonna have to come back to this later if ever. personal issues arose and i just havent had time to keep track of this, i keep forgetting i have this up.
discord log from my todo list for what to do next if i or anyone else picks this back up:

>Rohesie 06/03/2020
You need to move CAN_BE_HIT down to the atom level as well, @PKPenguin321 , and only enable it for objects and walls meant to be hit.

>PKPenguin321 06/03/2020
that's  gonna be rough because it's a flag baked into objects

>Rohesie 06/03/2020
((obj_flags & CAN_BE_HIT) && I.attack_obj(src, user)) was doing the trick before

>PKPenguin321 06/03/2020
on the same var as things like EMAGGED
maybe i could make it a resistance flag

>Rohesie 06/03/2020
It will be easier than you think, hold on.
Ah, that could work
Hmm, well, changing it to a resistance flag won't be trivial, no

>PKPenguin321 06/03/2020
i think it won't be trivial because there's some code about a mappers-only var
that touches CAN_BE_HIT
but correct me if i
am wrong

>Rohesie 06/03/2020
Is there room in flags_atom ?
Hmm? let me check

>PKPenguin321 06/03/2020
just curious, why would flags_atom be preferable to a resistance flag?

>Rohesie 06/03/2020
It actually wouldn't

>Rohesie 06/03/2020
And yes, set_obj_flags logic would have to be tweaked here
luckily set_obj_flags is only used for EMAGGED currently
not for CAN_BE_HIT as of now

**Comment by spookydonut** [comment]
@PKPenguin321 stay safe man good luck

**Pull request closed without merging**